DOC: SciPy 1.15.0 relnotes

tylerjereddy opened the pull request:
Early draft of SciPy `1.15.0` release notes. Skipping CI completely for now, though local render of the draft notes looks nice so far. Folks may continue to update the wiki or make suggestions here--either is fine--I simply start this because the initial transcription takes longer than you might think so getting it rolling helps avoid last minute pain.

TODO:

- [x] transcribe wiki version to RST, wrap at 80 char line lengths, fix links/errors, check render (last done: Nov 30th at 8:09 pm US mountain time)
- [x] check the wiki for updates again before branching and update this PR accordingly
- [x] manually scan through `ENH` PRs and add entries not already present in current notes (last done on Dec. 3/2024)
- [x] author list update (**repeat just before branching**)
- [x] issue/PR list update (**repeat just before branching**)
- [x] `.mailmap` update from gh-21940
- [x] gradually reactivate the CI here, when ready
- [x] finalize release notes highlights
- [x] Remove unused/empty (sub)sections in the release notes

tylerjereddy committed:
WIP, DOC: SciPy 1.15.0 relnotes

[skip ci] [ci skip] [skip cirrus] [skip circle]

j-bowhay commented on doc/source/release/1.15.0-notes.rst:
@adammj maybe you could suggest an entry for your improvement to lombscargle

adammj replied:
@j-bowhay Can do, I'll take a look at some examples and get back to you. Just asking to confirm, should https://github.com/scipy/scipy/pull/21729 go into 1.15.0?

adammj replied:
I think getting into the weeds about the algorithm change might be too verbose for the release notes (especially, since the output is the same).

Suggested entry:
- `scipy.signal.lombscargle` has two new arguments, ``weights`` and ``floating_mean``, enabling sample weighting and removal of an unknown y-offset independently for each frequency. Additionally, the ``normalize`` argument includes a new option to return the complex representation of the amplitude and phase.

tylerjereddy commented on doc/source/release/1.15.0-notes.rst:
```diff
@@ -64,16 +172,131 @@ New features
+  implementation of univariate, continuous distributions:
+
+  - Speed, accuracy, and memory advantages of the new infastructure are
+    detailed in
```
mostly note to self--this is the only "incomplete" sentence I noticed last night, probably because I transcribed the wiki release notes a bit early

mdhaber replied:
I think I left a blank to add a link to a tutorial when it merges.

tylerjereddy committed:
DOC: PR 21982 revisions

* Add a release note entry for `lombscargle` changes
based on reviewer feedback.

[skip ci] [ci skip] [skip cirrus] [skip circle]

rgommers commented on doc/source/release/1.15.0-notes.rst:
```diff
@@ -24,38 +24,151 @@ This release requires Python 3.10+ and NumPy 1.23.5 or greater.
+existing code from sparse matrix to sparse array:
+:ref:`migration_to_sparray`. Both ``sparse.linalg`` and ``sparse.csgraph``
+work with either sparse matrix or sparse array and work internally with
+sparse array.
```
+1. I'd also suggest n-D COO arrays as a separate highlight.

rgommers replied:
`scipy.differentiate` may also deserve a mention as a highlight? It's rare we add a new top-level submodule.

lucascolley replied:
I think @ollybritton plans to write something about `integrate.cubature` too, which can go in the highlights.

j-bowhay replied:
I would also put forward the new `optimize.elementwise` methods and `interpolate.AAA`  for rational approximation. This looks like the best release in a while for new features!

rgommers commented on doc/source/release/1.15.0-notes.rst:
```diff
@@ -87,11 +315,76 @@ Deprecated features
+    - When passed by position or legacy keyword, the behavior of the argument
+      will remain unchanged (for now).
+
+  - Beginning in SciPy ``1.17.0``:
```
I suggest to remove this, or reword briefly like _"It is planned that in 1.17.0 the legacy argument will start emitting warnings, and that in 1.19.0 the default behavior will change"_,  link to an issue where folks can see details and follow along. Plans for future releases often change, so this doesn't really belong in release notes.

rgommers reviewed (commented):
Thanks Tyler! Looks good overall.

For the first paragraph, I suggest changing `Python 3.10+` to `Python 3.10-3.13`, since that's more accurate and useful info.

Here is suggested content to describe free-threading support: https://github.com/rgommers/scipy/tree/relnotes-115-freethreading. Could you please pull that in @tylerjereddy?

I added a few other high-level comments; no detailed line by line review yet since this is draft.

mdhaber commented on doc/source/release/1.15.0-notes.rst:
```diff
@@ -24,38 +24,151 @@ This release requires Python 3.10+ and NumPy 1.23.5 or greater.
+work with either sparse matrix or sparse array and work internally with
+sparse array.
+
+A transition has started towards unifying the SciPy API with regards to
```
I was going to suggest just removing it. Only one function didn't already support `Generator`, IIRC.

jorenham commented on doc/source/release/1.15.0-notes.rst:
```suggestion
  ``bracket_root``, ``find_root``, ``bracket_minimum``, and ``find_minimum``
```

tylerjereddy committed:
WIP, DOC: PR 21982 revisions

* Update the release notes for SciPy `1.15.0` following
manual inspection of all 123 merged enhancement-labeled PRs with
this milestone.

* Add forgotten array API support section from the release notes
wiki to the RST release notes.

[skip ci] [ci skip] [skip cirrus] [skip circle]

j-bowhay commented on doc/source/release/1.15.0-notes.rst:
```diff
@@ -24,56 +24,353 @@ This release requires Python 3.10+ and NumPy 1.23.5 or greater.
+  evaluation of a convergent integral using tanh-sinh quadrature.
+- The new `scipy.integrate.cubature` function supports multidimensional
+  integration, has support for approximating integrals with
+  one or more sets of infinite limits, and supports custom integration rules.
```
This is not quite true (yet)
```suggestion
  one or more sets of infinite limits.
```

tylerjereddy commented:
Ok, I did the manual read through of all 123 ENH-labelled PRs that were merged under the `1.15.0` milestone and filled in some things that were missing. I cast a fairly wide net, adding a few things I thought were borderline, since I figured it would be easier to remove them than to have other also scanning through the pages of PRs.

I also added the array API support section--somehow I forgot that in the first pass at transcribing from the wiki page a few days ago.

Caution: I caught some deprecations that were not mentioned--particularly messy was stuff surrounding `special` like `assoc_legendre_p_all` -- I was about to add a release note saying that `lpn` et al. had a new keyword/improvement, and then when I looked at them, they were deprecated in the same release as the improvement!

jorenham commented on doc/source/release/1.15.0-notes.rst:
a cool thing that might be worth mentioning, is that these sample estimators are *unbiased* (which isn't the case for e.g. `np.std` or `scipy.stats.moment` in general)

dschult commented on doc/source/release/1.15.0-notes.rst:
```diff
@@ -24,56 +24,353 @@ This release requires Python 3.10+ and NumPy 1.23.5 or greater.
+:ref:`migration_to_sparray`. Both ``sparse.linalg`` and ``sparse.csgraph``
+work with either sparse matrix or sparse array and work internally with
+sparse array.
+
```
Here's a separate highlight for n-D COO arrays:
```suggestion

Sparse arrays now provide basic support for n-D arrays in the COO format including add, subtract, reshape, transpose, matmul, dot, tensordot and others. More functionality is coming in future releases.

```

tylerjereddy committed:
DOC: PR 21982 revisions

* Update the `.mailmap` file according to gh-21940, then
update the authors list in the release notes for SciPy `1.15.0`
accordingly.

[skip ci] [ci skip] [skip cirrus] [skip circle]

tylerjereddy committed:
DOC: PR 21982 revisions

* Add issue/PR lists to SciPy `1.15.0` release notes.

[skip ci] [ci skip] [skip cirrus] [skip circle]

tylerjereddy committed:
DOC: PR 21982 revisions

* Remove unused subsections in the SciPy `1.15.0` release notes.

[skip ci] [ci skip] [skip cirrus] [skip circle]

jorenham commented:
I'm also compiling a similar 1.15.0 checklist at https://github.com/jorenham/scipy-stubs/issues/250. So in case anyone is willing to see if I missed something, that would be certainly appreciated!

tylerjereddy committed:
DOC: PR 21982 revisions

* Address the first batch of reviewer comments related
to the draft SciPy `1.15.0` release notes.

[docs only]

tylerjereddy commented:
I've attempted to catch up with all the reviewer comments now a well--I'll leave them "unresolved" for a bit in case folks prefer to check that I've genuinely resolved their concern.

I'm also lifting the draft status and allowing the docs-only CI to flush through now since it looks pretty nice locally. Of course, this will remain informally WIP until branching proper.

tylerjereddy commented:
I'd also be "ok" with mentioning `scipy-stubs` in the "Other" section perhaps? Looks like Joren is pretty organized there and that's at least of peripheral interest to consumers.

jorenham commented:
> I'd also be "ok" with mentioning `scipy-stubs` in the "Other" section perhaps? 

That'd be awesome 😄. 

I'll try my best to release `scipy-stubs 1.15.0.0` *before* the final `scipy 1.15.0`. Assuming that there won't be any API changes after the RC has been released, then those 3-4 weeks should be more than enough time for me to achieve that 👌🏻.

jorenham commented:
> > I'd also be "ok" with mentioning `scipy-stubs` in the "Other" section perhaps?
> 
> That'd be awesome 😄.
> 
> I'll try my best to release `scipy-stubs 1.15.0.0` _before_ the final `scipy 1.15.0`. Assuming that there won't be any API changes after the RC has been released, then those 3-4 weeks should be more than enough time for me to achieve that 👌🏻.

So in case it wasn't clear yet; I don't plan on releasing a `scipy-stubs 1.15.0rc0.0`, and jump right to `1.15.0.0`. And even if it turns out that I messed up something, then `scipy-stubs` isn't bound to any release schedule, so I'll be able to fix it right away.

lucascolley commented on doc/source/release/1.15.0-notes.rst:
```suggestion
- All functions in the ``scipy.misc`` submodule have been removed.

```

jorenham replied:
so not the module itself 🤔 ?

jorenham replied:
IMHO I think a fail-fast approach would've been better; this way it'll inevitably result in a delayed `AttributeError` (unless the import is redundant), as opposed to an `ImportError` from the crime scene itself. 
Empty modules also can't be marked with `@decorated` like functions and classes can, so type-checkers won't be helpful as witnesses.

But what's done is done I suppose 🤷🏻

mdhaber replied:
This is still being discussed to some extent at https://github.com/scipy/scipy/issues/15765#issuecomment-2519263147.

lucascolley commented on doc/source/release/1.15.0-notes.rst:
```suggestion
- `signal.sosfreqz` has been renamed to `signal.freqz_sos`.
  New code should use the new name.
  The old name is maintained as an alias for backwards compatibility.
- Testing thread-safety improvements related to Python ``3.13t`` have been
  made in: `scipy.special`, `scipy.spatial`, `scipy.sparse`,
  `scipy.interpolate`.
```

lucascolley commented on doc/source/release/1.15.0-notes.rst:
@tylerjereddy prophetically, at least one bug remains in the `gh_lists` shenanigans :) we should not be escaping asterisks if they occur within code blocks. I can try to look at this before rc1.

jorenham commented on doc/source/release/1.15.0-notes.rst:
I believe [`dask`](https://github.com/dask/dask) and [`sparse`](https://github.com/pydata/sparse) also have (some) array-api support

lucascolley replied:
Dask we know has issues (gh-20956), and `sparse` we haven't really taken a look at yet. So I think we'd just be encouraging bug reports which may be inevitable or not actionable right now.

jorenham commented on doc/source/release/1.15.0-notes.rst:
I believe that `distance.kulczynski1` and `distance.sokalmichener` were deprecated in #21572

jorenham replied:
this can be resolved now (I have no super cow powers here)

jorenham commented on doc/source/release/1.15.0-notes.rst:
`toeplitz` doesn't support batching (yet): it currently ravels >1-d input, which has been deprecated, so it can support batching in 1.17

jorenham replied:
https://github.com/scipy/scipy/pull/21446/files#diff-6482beada610f60c724811cf77561f2bde8fd44ed2349939e505d28746a6d078R38-R42

mdhaber replied:
Oops. Good catch.

jorenham commented on doc/source/release/1.15.0-notes.rst:
```diff
@@ -17,96 +17,1431 @@ run your code with ``python -Wd`` and check for ``DeprecationWarning`` s).
 1.15.x branch, and on adding new features on the main branch.
 
-This release requires Python 3.10+ and NumPy 1.23.5 or greater.
+This release requires Python 3.10-3.13 and NumPy 1.23.5 or greater.
```
Oh and I see that SPEC 0 also recommends dropping support for Python 3.10 in Q4 2024.

mdhaber replied:
Re: Python version, it sounds fine to stay in sync with NumPy as long as they don't get too far behind.

Re: Python versions, it will be very helpful when the minimum NumPy version is 2.0+.

Note that we will actually be past the SPEC 0 support window for NumPy 1.25 when SciPy 1.15 is released, but we tend to lag behind by a version or so, perhaps to avoid ending up on the borderline.

If we do NumPy 1.24 for 1.15 (Jan 2025), NumPy 1.25 for SciPy 1.16 (June 2025), and NumPy 1.26 for SciPy 1.17 (Jan 2026), we could probably skip to NumPy 2.1 for SciPy 1.18 (Jun 2026).

tylerjereddy committed:
DOC: PR 21982 revisions

* Mention the `scipy-stubs` package in the
release notes.

* Mention `ndonnx` in the array API support section.

* Mention the removal of the functions in `scipy.misc` and the
renaming of `sosfreqz` to `freqz_sos`.

* Remove an undesirable escape character from the release notes
formatting for issue 21044.

* Mention the deprecations of `distance.kulczynski1` and
`distance.sokalmichener`.

* Clarified the current status of batching support for
`scipy.linalg.toeplitz`.

tylerjereddy commented:
I pushed in a commit that aims to address the latest round of reviewer comments. Python `3.10` will be supported by the upcoming NumPy release, so I suspect we may want to drop `3.10` in sync with them (for the release after that). It may cause a bit of last-minute chaos to bump the lowest NumPy version, maybe do it just after branching?

jorenham commented on doc/source/release/1.15.0-notes.rst:
```diff
@@ -407,6 +413,8 @@ Backwards incompatible changes
 *************
 Other changes
 *************
+- A separate accompanying type stubs package, ``scipy-stubs``, will be made
+  available with the ``1.15.0`` release.
```
I suppose that installation instructions might help here. I've documented them at https://github.com/jorenham/scipy-stubs?tab=readme-ov-file#installation.

Now that I think about, maybe we could add it as an "extra", so that it's simply `pip scipy[stubs]`? I'm not sure what the alternative for this is for conda though, but `scipy-stubs` is available through `conda-forge` already 🤔

lucascolley replied:
it works quite nicely on conda-forge, `pixi add scipy` gives you SciPy without the tests and `pixi add scipy-tests` gives you SciPy with the tests - see https://github.com/conda-forge/scipy-feedstock/blob/1a8ff378bebd3f0b0bc6a2d1662a77d194ac59a1/recipe/meta.yaml#L205. It seems feasible to me to make something like `pixi add scipy-with-stubs` available with a runtime dependency on `scipy-stubs`.

mdhaber commented on doc/source/release/1.15.0-notes.rst:
```suggestion
  been deprecated. It will support batching in SciPy ``1.17.0``.
  - The ``random_state`` and ``permutations`` arguments of `scipy.stats.ttest_ind`
     are deprecated. Use ``method`` to perform a permutation test.
```

tylerjereddy committed:
DOC: PR 21982 revisions

* Mention argument deprecations in `ttest_ind` in the
SciPy `1.15.0` release notes.

* A link to the `scipy-stubs` installation instructions has
been added to the SciPy `1.15.0` release notes.

[docs only]

tylerjereddy commented:
I included Matt's addition of `ttest_ind`-related argument deprecations and a hyperlink to Joren's install instructions section for `scipy-stubs` (I figure even if the instructions change a bit per above, Joren could always just update that section in the stubs repo a bit).

tylerjereddy commented:
@mdhaber `stats` has been moving pretty fast last few days--are we "green light" form your side to branch around tomorrow?

@h-vetinari how about you? any concerns on your end/conda-forge side?

tylerjereddy commented:
I believe the only unresolved release notes matters at the moment are:

1) a potential tutorial/link for new `stats` distribution infra 
2) Matt does seem to have an interest in us bumping up to `1.24.4` minimum supported NumPy version for this release cycle (@rgommers is that "ok" with you?). My main objection above was mostly just that this would be a little annoying to do this close to branching, but it can certainly be done (probably would miss a few subtle spots for clean up).

mdhaber commented:
We don't necessarily have to update min NumPy before branching, but I remember that there was also a preference against updating min NumPy by two versions in one SciPy version. It just seems that if we miss NumPy version bumps and never do two in one SciPy version, we will gradually fall behind the SPEC 0 schedule.

rgommers commented:
Ah seeing this dependency version discussion after commenting on gh-22012. I don't have much more to add to the comment I posted there, other than that there is no rule that says we can't drop 2 versions at once (and IIRC we've done so at least once before).

tylerjereddy commented:
Alright, I think we're on the same page, branching is imminent then I think (a few hours probably, just 2 other PRs with the milestone left, likely to be bumped).

tylerjereddy committed:
DOC: PR 21982 revisions

* Deal with gh-22020.

* Update the author list one more time in the
SciPy `1.15.0` release notes, before branching.

* Update the issue/PR lists before branching,
and update the release notes again for ENH PRs
that were recently merged.

* Remove the tentative mention of docs relating
to new stats distn infra for now.

[docs only]

jorenham commented on doc/source/release/1.15.0-notes.rst:
#21746 added `scipy.stats.{exp,log}`

jorenham replied:
#21752 added `truncate`, `abs`, and `Mixture`

mdhaber replied:
https://github.com/scipy/scipy/pull/21982/files#r1874575198

tylerjereddy commented:
Ok, I'm doing some fairly fast updating of the relnotes from recently-merged PRs locally for now..

tylerjereddy commented:
I've pushed in the latest revisions/relnotes additions--we'll see if the docs build in CI agrees with the fix for gh-22020.

jorenham commented on doc/source/release/1.15.0-notes.rst:
```diff
@@ -291,6 +292,17 @@ and support several Array API compatible array libraries in addition to NumPy
+  take the absolute value of a random variable. The new ``FoldedDistribution``
+  represents the distribution underlying the absolute value of a random
+  variable.
+- Added ``OrderStatisticDistribution``, a class representing the distribution
+  of order statistics of samples drawn from another distribution.
```
`OrderStatisticDistribution` isn't public, and the `scipy.stats.order_statistic` factory function is what should be used for this:
https://github.com/scipy/scipy/pull/21790/files

mdhaber replied:
```suggestion
```
I'll add this (`order_statistic`) up above.

jorenham commented on doc/source/release/1.15.0-notes.rst:
```diff
@@ -291,6 +292,17 @@ and support several Array API compatible array libraries in addition to NumPy
+  other than ``(2, 2)``.
+- `scipy.stats.truncate` adds the ability to truncate the distribution
+  underlying a random variable and `scipy.stats.abs` provides the ability to
+  take the absolute value of a random variable. The new ``FoldedDistribution``
```
I don't think the `FoldedDistribution` is exposed in `scipy.stats`, or am I wrong here, @mdhaber ?

mdhaber replied:
You're right, thanks! I will read through this again now.

mdhaber commented on doc/source/release/1.15.0-notes.rst:
```suggestion
```

mdhaber commented on doc/source/release/1.15.0-notes.rst:
```suggestion
```
This is only part of the `sample` method of the new RV objects, so it's not really an enhancement that needs to be covered separately.

mdhaber commented on doc/source/release/1.15.0-notes.rst:
```suggestion
```

mdhaber commented on doc/source/release/1.15.0-notes.rst:
```suggestion
- `scipy.stats.Mixture` has been added to represent mixture distributions.
- Instances of `scipy.stats.Normal` and the classes returned by
  `scipy.stats.make_distribution` are supported by several new mathematical
  transformations.
  - `scipy.stats.truncate` for truncation of the support.
  - `scipy.stats.order_statistic` for the order statistics of a given number of IID
    random variables.
  - `scipy.stats.abs`, `scipy.stats.exp`, and `scipy.stats.log`. For example,
    ``scipy.stats.abs(Normal())`` is distributed according to the folded normal
    and ``scipy.stats.exp(Normal())`` is lognormally distributed.
```

mdhaber commented on doc/source/release/1.15.0-notes.rst:
```suggestion
    interface. In typical cases, its methods are faster and more accurate than
```

mdhaber reviewed (approved):
I think this will LGTM after these last few suggestions.

h-vetinari commented:
> @h-vetinari how about you? any concerns on your end/conda-forge side?

I started a testrun based on current `main` in https://github.com/conda-forge/scipy-feedstock/pull/284, and it looks like everything is passing (emulated tests are still running, but not expecting any issues there)

tylerjereddy committed:
DOC: PR 21982 revisions

* Apply final round of suggestion from code review
of SciPy `1.15.0` release notes.

[docs only]

tylerjereddy commented:
Alright, thanks Matt and Joren for helping with last-minute cleanups. I think got most of the recent round of comments dealt with now and if the docs CI passes again I'll likely squash merge and branch.

h-vetinari commented:
> > @h-vetinari how about you? any concerns on your end/conda-forge side?
> 
> I started a testrun based on current `main` in [conda-forge/scipy-feedstock#284](https://github.com/conda-forge/scipy-feedstock/pull/284), and it looks like everything is passing (emulated tests are still running, but not expecting any issues there)

Quick update: I opened https://github.com/scipy/scipy/issues/22022 (non-critical), and there's some new timeouts in emulation `test_quad_vec_pool` and `test_quad_vec_pool_args`. 

More interestingly, after working around a conda-build issue, it also looks like we'll be able to add free-threading builds for scipy right away (it builds & test suite passes 🥳).
